src/java.base/share/classes/java/util/concurrent/ForkJoinWorkerThread.java

```diff
@@ -225,6 +226,25 @@ static boolean hasKnownQueuedWork() {
                   (sq = qs[i]) != null && sq.top - sq.base > 0) ||
                  q.top - q.base > 0));
     }
+
+    /**
+     * Clears ThreadLocals, and if necessary resets ContextClassLoader
+     */
+     final void resetThreadLocals() {
+         if (U.getReference(this, THREADLOCALS) != null)
+             U.putReference(this, THREADLOCALS, null);
+         if (U.getReference(this, INHERITABLETHREADLOCALS) != null)
+             U.putReference(this, INHERITABLETHREADLOCALS, null);
+         if ((this instanceof InnocuousForkJoinWorkerThread) &&
+             ((InnocuousForkJoinWorkerThread)this).needCCLReset())
+             super.setContextClassLoader(ClassLoader.getSystemClassLoader());
+     }
+
+    private static final Unsafe U = Unsafe.getUnsafe();
+    private static final long THREADLOCALS
+        = U.objectFieldOffset(Thread.class, "threadLocals");
+    private static final long INHERITABLETHREADLOCALS
+        = U.objectFieldOffset(Thread.class, "inheritableThreadLocals");
```

Comment:
@AlanBateman Thoughts here? 🤔

Reply:
resetThreadLocals looks good although. A discussion point is whether it the reset should be done for all FJP instances, not just the common pool but not this PR.

Reply:
Thanks Alan. Yes, I think it should be considered to do it for all FJP instances.

Reply:
Well, almost all. There is a protected FJWT constructor with explicit arg about preserving ThreadLocals,, in which case it should also not reset CCL. (Also, I just noticed that the Thread ctor with !inherit now also forces using System classLoader, so the FJWT ctor no longer needs to in this case.) Will do as part of SM-removal pr coming next.